Non-breaking white spaces characters not checked for correctly in Scroll To Text Fragment directive search.

megangardner opened the pull request:
#### c4e75098c6f269ae0a48121e11ca09a1ab7a4cf1
<pre>
Non-breaking white spaces characters not checked for correctly in Scroll To Text Fragment directive search.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244231">https://bugs.webkit.org/show_bug.cgi?id=244231</a>

Reviewed by Darin Adler and Yusuke Suzuki.

We were using isSpaceOrNewline to check if a character was a space or new line, but this
does not check for non-breaking white spaces. We really need to use the ICU function that
checks all defined white space characters in Unicode. Add a function to encapsulate that
check. It&apos;s possible that other instances of isSpaceOrNewline should actually be isUnicodeWhitespace
so we have filed a radar to check other call-sites to see if there are other places we should switch
to this new function.

Updated results for content-visibility-048 because the addition of the scroll to text fragment resources
changed the failure.

* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/find-range-from-text-directive-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/find-range-from-text-directive-target.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/find-range-from-text-directive.html: Added.
* LayoutTests/tests-options.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-048-expected.txt:
* Source/WTF/wtf/text/StringImpl.h:
(WTF::isUnicodeWhitespace):
* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::advanceRangeStartToNextNonWhitespace):

Canonical link: <a href="https://commits.webkit.org/253737@main">https://commits.webkit.org/253737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22568ce113b5a93bed810538e7b45fd8d53f1efa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86945 "failed 1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17812 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95817 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149531 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90924 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29399 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79089 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90995 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92561 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23738 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/23751 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/66758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78852 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27140 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72514 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27075 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/13877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25855 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2649 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28755 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75299 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28699 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/33154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16658 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
isUnicodeWhitespace part looks good to me. It is possible that we can further optimize the branching based on the type of input `character`. (For LChar, `isASCIISpace(character) || character == noBreakSpace` would be enough, but, we can just place it as FIXME).

megangardner committed:
Non-breaking white spaces characters not checked for correctly in Scroll To Text Fragment directive search.
https://bugs.webkit.org/show_bug.cgi?id=244231

Reviewed by Darin Adler and Yusuke Suzuki.

We were using isSpaceOrNewline to check if a character was a space or new line, but this
does not check for non-breaking white spaces. We really need to use the ICU function that
checks all defined white space characters in Unicode. Add a function to encapsulate that
check. It's possible that other instances of isSpaceOrNewline should actually be isUnicodeWhitespace
so we have filed a radar to check other call-sites to see if there are other places we should switch
to this new function.

Updated results for content-visibility-048 because the addition of the scroll to text fragment resources
changed the failure.

* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/find-range-from-text-directive-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/find-range-from-text-directive-target.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/find-range-from-text-directive.html: Added.
* LayoutTests/tests-options.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-048-expected.txt:
* Source/WTF/wtf/text/StringImpl.h:
(WTF::isUnicodeWhitespace):
* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::advanceRangeStartToNextNonWhitespace):

Canonical link: https://commits.webkit.org/253737@main
